ares/n64/pi/bus.hpp

```diff
@@ -0,0 +1,79 @@
+template<u32 Size>
+inline auto PI::busRead(u32 address) -> u64 {
+  static constexpr u64 unmapped = 0;
+
+  if (unlikely(io.ioBusy)) {
+    io.ioBusy = 0; //technically, we should wait until Queue::PI_BUS_Write
+    if constexpr(Size == Byte) {
+      return io.busLatch >> ((~address & 3) * 8);
+    }
+    if constexpr(Size == Half) {
+      return io.busLatch >> ((~address & 1) * 16);
+    }
+    if constexpr(Size == Word) {
+      return io.busLatch;
+    }
+    if constexpr(Size == Dual) {
+      return io.busLatch;
+    }
+  }
+
+  if(address <= 0x0500'03ff) return dd.c2s.read<Size>(address);
+  if(address <= 0x0500'04ff) return dd.ds.read<Size>(address);
+  if(address <= 0x0500'057f) return dd.read<Size>(address);
+  if(address <= 0x0500'05bf) return dd.ms.read<Size>(address);
+  if(address <= 0x05ff'ffff) return unmapped;
+  if(address <= 0x063f'ffff) return dd.iplrom.read<Size>(address);
+  if(address <= 0x07ff'ffff) return unmapped;
+  if(address <= 0x0fff'ffff) {
+    if(cartridge.ram  ) return cartridge.ram.read<Size>(address);
+    if(cartridge.flash) return cartridge.flash.read<Size>(address);
+    return unmapped;
+  }
+  if(address >= 0x13ff'0000 && address <= 0x13ff'ffff) {
+    return cartridge.isviewer.read<Size>(address);
+  }
+  return cartridge.rom.read<Size>(address);
+}
+
+template<u32 Size>
+inline auto PI::busWrite(u32 address, u64 data) -> void {
+  if(io.ioBusy) return;
+  io.ioBusy = 1;
+  if constexpr(Size == Byte) {
+    io.busLatch = data << ((~address & 3) * 8);
+  }
+  if constexpr(Size == Half) {
+    io.busLatch = data << ((~address & 1) * 16);
+  }
+  if constexpr(Size == Word) {
+    io.busLatch = data;
+  }
+  if constexpr(Size == Dual) {
+    io.busLatch = data >> 32;
+  }
+
+  queue.insert(Queue::PI_BUS_Write, 300);
```

Comment:
Are we at risk of overflowing the queue? I'm concerned about DMA in particular, as it goes through this code path.

Reply:
In general, only one write can be ongoing at the same time. There are two shortcuts right now (reading after writing, and isviewer) so I guess I could refactor those into a write abort functions that also removes the queue entry, so we avoid the problem altogether